### PR TITLE
UI Pagination

### PR DESCRIPTION
**Pull request by nhaagen**
Applies to all kind of data-lists where the shown records are limited; use a pagination-control to digest the data in chunks.

Rules are [here](https://github.com/ILIAS-eLearning/ILIAS/pull/671/files?diff=unified#diff-8a479116309e579f1e1c5fbec91dad6a).

bildschirmfoto_2017-10-24_16-23-21

bildschirmfoto_2017-10-24_16-23-56

bildschirmfoto_2017-10-24_16-24-13

bildschirmfoto_2017-10-24_16-24-39

### CODE REVIEW

File: src/UI/Implementation/Component/ViewControl/Pagination.php
Comment (klees) [inline src/UI/Implementation/Component/ViewControl/Pagination.php:93]: This is missing on the interface.

File: tests/UI/Component/ViewControl/PaginationTest.php
Comment (klees) [inline tests/UI/Component/ViewControl/PaginationTest.php:13]: Please add a test for the rendering at some point.

File: tests/UI/Component/ViewControl/PaginationTest.php
Comment (klees) [inline tests/UI/Component/ViewControl/PaginationTest.php]: paramter -> parameter

File: src/UI/Component/ViewControl/Factory.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Factory.php]: Why the underlines?

File: src/UI/Component/ViewControl/Factory.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Factory.php]: There is more to that. One of those buttons is always active right? The ones on the left and right look special, what are those? What about the pagination of empty lists? Is the Pagination also empty?

File: src/UI/Component/ViewControl/Factory.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Factory.php]: What happens with the pagination itself? Which button will be active? What happens if the one on the left or right is clicked? Are left and right deactivated if there is no next?

File: src/UI/Component/ViewControl/Factory.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Factory.php]: On top or bottom? If both are possible, suggest that with a MAY rule.

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: This is far to much JS code placed in PHP. Factor this out to a JS file add call the according function with the necessary params only here. In PHP we only want JS-One-liners.

File: src/UI/Component/ViewControl/Pagination.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Pagination.php]: The effect of this function was only clear to me after looking at the examples. I think there is some double usage of the word "Entries" here. Above we used entries as label for the item in the controlled data list. Here we use it the name the numbers in the pagination itself. Maybe rename to "withMaxPaginiationButtons" or similar.

**Review by Amstutz — CHANGES_REQUESTED** [review]
Interface:
-> The function of the Previous/Next Glyph on the right and left is not really clear from the description.
-> We also need some Button/Glyph allowing to jump to the first and last of the pages (easy to see in the examples with lots of entries).
-> Some details (see comments)

Implementation:
-> Before merging, a more complete set of tests is needed (especially some rendering test).

**Comment by nhaagen** [comment]
Thanks for your comments, @Amstutz. I think, they were all adressed. Please have another go.

**Comment by Amstutz** [comment]
@nhaagen Thanks a lot for the improvements. To me this is JF ready. While presenting, please layout a clear strategy on what elements in 5.4 this control could be used (are there more related controls to come? I hope so).

**Review by Amstutz — APPROVED** [review]

File: src/UI/Component/ViewControl/Factory.php
Comment (atoedt) [inline src/UI/Component/ViewControl/Factory.php]: Pagination allows structured data being displayed in chunks by limiting the number of entries shown. Pagination provides the user with controls to leaf through chunks of entries.

File: src/UI/Component/ViewControl/Factory.php
Comment (atoedt) [inline src/UI/Component/ViewControl/Factory.php]: Pagination is a collection of shy-buttons to access individual chunks of entries, framed by additional shy-buttons with chevron left/right glyphs. 

Why: Do not use page, this is a difficult and blurry concept

File: src/UI/Component/ViewControl/Factory.php
Comment (atoedt) [inline src/UI/Component/ViewControl/Factory.php]: Clicking on a chunk-option navigates to the indicated chunk and the respective entries are displayed. The active option is rendered as an unavailable shy-button.
Clicking the chevron-buttons navigates to the previous (respectively: the next) chunk of entries. If no previous/next chunk is available, the chevron-button is rendered unavailable. 
If the pagination is used with a maximum of chunk-options to be shown, both first and last options are always displayed.

File: src/UI/Component/ViewControl/Factory.php
Comment (atoedt) [inline src/UI/Component/ViewControl/Factory.php]: The use of Pagination MUST only be used for structured data, like tables and lists.

File: src/UI/Component/ViewControl/Pagination.php
Comment (klees) [inline src/UI/Component/ViewControl/Pagination.php:111]: Do you see any chance to control this via some css or client side code instead of setting it here? It seems at if the value one wants to set here depends on some css-induced circumstances (font, paddings, ...) that the developer introducing that element won't know in advance. Another option would be to just fix this and remove this setter. I just cannot see how a developer using this element could sensibly control this.

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (klees) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: This looks rather complicated, maybe factor out some methods. This will especially help the mathy parts of that method.

**Review by klees — COMMENTED** [review]
LGTM in general. I really would like to have a solution for that "max-pagination-buttons"-problem. Some refactoring in the renderer would make this totally nice. Thx!

File: src/UI/Component/ViewControl/Pagination.php
Comment (nhaagen) [reply src/UI/Component/ViewControl/Pagination.php:111]: Imo, this is more of a usability decision than a visual thing. I like the option of limiting/unlimiting the pagination, since there might be numbered lists with easy-to-grasp steps, where you explicitly favor quick navigation over cleanliness.
Thus, you would still need a setter. Why not with a value, then?
However, where would you see the fix value? I personally think 5 is too low and 10 is too big, so I would go for 7, if at all.

**Comment by atoedt** [comment]
pagination_letters_users
pagination_letters_terms
These are two existing other pagination uses. Which are not covered by the article and possibly rightly so.  What do you think?

**Comment by nhaagen** [comment]
Thank you for the comments, @atoedt. I agree that the given examples are not actually paginations but indexes, although very close in visualization.
I'll update the rules with your suggestions.

File: src/UI/Component/ViewControl/Factory.php
Comment (klees) [inline src/UI/Component/ViewControl/Factory.php]: To much "NOT" in this line: "A Pagination MUST -NOT- only be used for structured data, like tables and lists."

**Comment by matthiaskunkel** [comment]
Jour Fixe, Oct 23, 2017: We appreciate this pull request. But we would like to include a dropdown for jumping quick to a page (like existing already in tables). Such a dropdown should only be offered in tables/lists with a big number of elements. It is up to the developer to decide when the dropdown is shown. Please take into consideration that this UI element needs to be handled on small devices, too (e.g. mobile phone). PR can be finalised.

**Comment by Amstutz** [comment]
What is the status of this? Is this ready to be reviewed again for merging?

**Comment by klees** [comment]
Nope, this is currently missing adjustments according to JF feedback.

**Comment by klees** [comment]
Hi @Amstutz,
any objections to merge this? Will do it manually, though, to solve the conflicts. Will also shorten the one long line in the rules then.
Regards!

File: src/UI/Component/ViewControl/Factory.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Factory.php]: chevron left/right are not valid glyph names. They are called Back and Next. Note that we do not use bootstrap terminology here, since the KS needs to survive the change from e.g. Bootstrap 3 to 4.

File: src/UI/Component/ViewControl/Pagination.php
Comment (Amstutz) [inline src/UI/Component/ViewControl/Pagination.php:129]: +1

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php:280]: wow, this is actually a quite nice way of doing that. Did not know about this up to know. However, is this already centralized anywhere. Looks kind of like it should be (e.g. src/HTTP).

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: Argl. terrible naming.

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: Why are no glyphs used here?

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: Those are Back/Next Glyphs.

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: Again why is are the Glyphs not used?

File: src/UI/Implementation/Component/ViewControl/Renderer.php
Comment (Amstutz) [inline src/UI/Implementation/Component/ViewControl/Renderer.php]: I hate this "Rocker" Termin, even in the variables.

File: src/UI/templates/default/ViewControl/tpl.pagination.html
Comment (Amstutz) [inline src/UI/templates/default/ViewControl/tpl.pagination.html]: Again, those Rockers...

File: src/UI/templates/js/ViewControl/pagination.js
Comment (Amstutz) [inline src/UI/templates/js/ViewControl/pagination.js]: What about those Todos?

**Review by Amstutz — CHANGES_REQUESTED** [review]
LGTM, almost there. Most importantly we have some Glyph (naming) issue. Why is the Glyph class not used here. I think this is a MUST. Further please add a Unit Test for the dropdown feature.

**Comment by Amstutz** [comment]
Any Update on this? I think this is really close to be able to be merged...

File: src/UI/Component/ViewControl/Pagination.php
Comment (klees) [reply src/UI/Component/ViewControl/Pagination.php:111]: Just leave it as it is.

**Comment by klees** [comment]
@Amstutz Thx for the reminder.

**Comment by klees** [comment]
@Amstutz May we merge this?

**Comment by klees** [comment]
@Amstutz I know you are afk for some time, please consider this as a friendly reminder to have a look here. Thx!

**Review by Amstutz — APPROVED** [review]
Thanks a lot @nhaagen. This looks good to me. 

Please clean up any merge conflicts, run the tests a final time and then this can be merged directly. 

This does not directly concern this PR, but @klees : If any of you have some time, it would be a good idea to get a feedback from someone with some designing background on the looks of those controls (e.g. Christoph). The interface is fine, and they work, however the looks can be improved. 

@nhaagen: If some feedback on the design on some component (e.g. the dropdowns) does not make sense to be implemented directly on the component itself, but is rather context specific, you might also introduce this on the less of the context (e.g. paginiation). We should do that with care, to not break the promise of consistency and adaptability, but it is not forbidden to do such things.

**Comment by klees** [comment]
@Amstutz I can send CK to have a look, but maybe @jbraun-ilias as CSS-maintainer want to take the lead here.

**Pull request merged**
